code/obj/zoldorf.dm

```diff
@@ -85,6 +40,113 @@ var/global/list/datum/zoldorfitem/zoldorf_items = list()
+		for (var/product in concrete_typesof(/datum/zoldorfitem/credit))
+			src.credit_items += new product
+
+	ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+		. = ..()
+		if(.)
+			return
+		switch(action)
+			if("returncash")
+				if(src.credits <= 0)
+					return
+				var/obj/item/moneyreturn = new /obj/item/currency/spacecash(get_turf(src),src.credits)
+				src.credits = 0
+				ui.user.put_in_hand_or_drop(moneyreturn)
+				return TRUE
+			if("soul_purchase")
+				var/item_name = params["item"]
+				if (!item_name)
+					return
+				var/datum/zoldorfitem/soul/purchasing = null
+				for(var/datum/zoldorfitem/soul/product in src.soul_items)
+					if(product.name == item_name)
+						purchasing = product
+				if (!istype(purchasing, /datum/zoldorfitem/soul))
+					return
+				if ((purchasing.stock <= 0) && !purchasing.infinite)
+					return
+				var/confirm = tgui_alert(ui.user, "Are you sure you want to sell [purchasing.soul_percentage]% of your soul?", "Confirm Transaction", list("Yes", "No"))
+				if(confirm != "Yes")
+					return
+				if (!soul_cost_check(ui.user, purchasing.soul_percentage))
+					return
+				var/mob/living/carbon/human/H = ui.user
+				if(H.sell_soul(purchasing.soul_percentage))
+					purchasing.on_bought(H)
+					purchasing.stock -= 1
+					src.partialsouls += purchasing.soul_percentage
+					src.updatejar()
+				return TRUE
+			if("credit_purchase")
+				var/item_name = params["item"]
+				if (!item_name)
+					return
+				var/datum/zoldorfitem/credit/purchasing = null
+				for(var/datum/zoldorfitem/credit/product in src.credit_items)
+					if(product.name == item_name)
+						purchasing = product
+				if(!istype(purchasing, /datum/zoldorfitem/credit))
+					return
+				if ((purchasing.stock <= 0) && !purchasing.infinite)
+					return
+				if(purchasing.price > src.credits)
+					boutput(ui.user, SPAN_ALERT("[src.name] stares blankly into your soul...begging you for more credits..."))
+					return
+				src.credits -= purchasing.price
+				purchasing.stock -= 1
+				purchasing.on_bought(ui.user)
+				return TRUE
+
+	ui_interact(mob/user, datum/tgui/ui)
+		ui = tgui_process.try_update_ui(user, src, ui)
+		if(!ui)
+			ui = new(user, src, "ZoldorfPlayerShop", src.name)
+			ui.open()
+
+	ui_data(mob/user)
+		. = ..()
+
+		.["credits"] = src.credits
+		.["user_soul"] = user.mind?.soul ? user.mind.soul : 0
+
+		var/list/products = list()
+		for(var/datum/zoldorfitem/soul/product as anything in src.soul_items)
+			products += list(list(
+				"name" = product.name,
+				"stock" = product.stock,
+				"infinite" = product.infinite,
+				"img" = product.img,
+				"soul_percentage" = product.soul_percentage,
+			))
+		for(var/datum/zoldorfitem/credit/product as anything in src.credit_items)
+			products += list(list(
+				"name" = product.name,
+				"stock" = product.stock,
+				"infinite" = product.infinite,
+				"img" = product.img,
+				"price" = product.price,
+			))
+		.["products"] = products
```

Comment:
It's a decently finite list and quite a niche object so this shouldn't be a massive issue, but if possible I would look to move the products list generation to `ui_static_data` and then call the relevant update proc whenever stock changes.

Reply:
Moved all product data to static and called `src.update_static_data(ui.user)` whenever the product count changes. credit/soul% are still in ui_data.